Sources/PIALibrary/Account/Data/RefreshAuthTokensChecker.swift

```diff
@@ -0,0 +1,110 @@
+
+import Foundation
+
+protocol RefreshAuthTokensCheckerType {
+    typealias Completion = ((NetworkRequestError?) -> Void)
+    func refreshIfNeeded(completion: @escaping Completion)
+}
+
+class RefreshAuthTokensChecker: RefreshAuthTokensCheckerType {
+    
+    let apiTokenProvider: APITokenProviderType
+    let vpnTokenProvier: VpnTokenProviderType
+    let refreshAPITokenUseCase: RefreshAPITokenUseCaseType
+    let refreshVpnTokenUseCase: RefreshVpnTokenUseCaseType
+    
+    // Number of days remaining before refreshing a token
+    private let daysUntilRefresh: Double = 30
+    
+    init(apiTokenProvider: APITokenProviderType, vpnTokenProvier: VpnTokenProviderType, refreshAPITokenUseCase: RefreshAPITokenUseCaseType, refreshVpnTokenUseCase: RefreshVpnTokenUseCaseType) {
+        self.apiTokenProvider = apiTokenProvider
+        self.vpnTokenProvier = vpnTokenProvier
+        self.refreshAPITokenUseCase = refreshAPITokenUseCase
+        self.refreshVpnTokenUseCase = refreshVpnTokenUseCase
+    }
+    
+    func refreshIfNeeded(completion: @escaping Completion) {
+        
+        switch (shouldRefreshApiToken(), shouldRefreshVpnToken()) {
+        case (true, true):
+            refreshBothTokens(with: completion)
+        case (true, false):
+            refreshApiToken(with: completion)
+        case(false, true):
+            refreshVpnToken(with: completion)
+        case(false, false):
+            completion(nil)
+        }
+    }
+    
+}
+
+// MARK: - Refresh tokens helpers
+
+private extension RefreshAuthTokensChecker {
+    
+    func refreshBothTokens(with completion: @escaping Completion) {
+        refreshAPITokenUseCase() { refreshApiTokenError in
+            if let refreshApiTokenError {
+                completion(refreshApiTokenError)
+            } else {
+                self.refreshVpnTokenUseCase() { refreshVpnTokenError in
+                    if let refreshVpnTokenError {
+                        completion(refreshVpnTokenError)
+                    } else {
+                        completion(nil)
+                    }
+                }
+            }
+        }
+    }
+    
+    func refreshApiToken(with completion: @escaping Completion) {
+        refreshAPITokenUseCase() { error in
+            self.handleRefreshResponse(with: error, completion: completion)
+        }
+    }
+    
+    func refreshVpnToken(with completion: @escaping Completion) {
+        refreshVpnTokenUseCase() { error in
+            self.handleRefreshResponse(with: error, completion: completion)
+        }
+    }
+    
+    func handleRefreshResponse(with error: NetworkRequestError?, completion: @escaping Completion) {
+        if let error {
+            completion(error)
+        } else {
+            completion(nil)
```

Comment:
nit: wouldn't this one line `completion(error)` achieve the same given that its parameter is either an error object or null? And if that's the case, we might not need the method altogether 🤔

Reply:
yes, absolutely , you're right 👍